extern/storage-sealing/states_replica_update.go

```diff
@@ -0,0 +1,184 @@
+	if sector.UpdateSealed == nil || sector.UpdateUnsealed == nil {
+		return xerrors.Errorf("invalid sector %d with nil UpdateSealed or UpdateUnsealed output", sector.SectorNumber)
+	}
+	if sector.CommR == nil {
+		return xerrors.Errorf("invalid sector %d with nil CommR", sector.SectorNumber)
+	}
+	vanillaProofs, err := m.sealer.ProveReplicaUpdate1(sector.sealingCtx(ctx.Context()), m.minerSector(sector.SectorType, sector.SectorNumber), *sector.CommR, *sector.UpdateSealed, *sector.UpdateUnsealed)
+	if err != nil {
+		return ctx.Send(SectorProveReplicaUpdate1Failed{xerrors.Errorf("prove replica update (1) failed: %w", err)})
+	}
+	return ctx.Send(SectorProveReplicaUpdate1{
+		Out: vanillaProofs,
+	})
+}
+
+func (m *Sealing) handleProveReplicaUpdate2(ctx statemachine.Context, sector SectorInfo) error {
+	if sector.UpdateSealed == nil || sector.UpdateUnsealed == nil {
+		return xerrors.Errorf("invalid sector %d with nil UpdateSealed or UpdateUnsealed output", sector.SectorNumber)
+	}
+	if sector.CommR == nil {
+		return xerrors.Errorf("invalid sector %d with nil CommR", sector.SectorNumber)
+	}
+	if sector.ProveReplicaUpdate1Out == nil {
+		return xerrors.Errorf("invalid sector %d with nil ProveReplicaUpdate1 output", sector.SectorNumber)
+	}
+	proof, err := m.sealer.ProveReplicaUpdate2(sector.sealingCtx(ctx.Context()), m.minerSector(sector.SectorType, sector.SectorNumber), *sector.CommR, *sector.UpdateSealed, *sector.UpdateUnsealed, sector.ProveReplicaUpdate1Out)
+	if err != nil {
+		return ctx.Send(SectorProveReplicaUpdate2Failed{xerrors.Errorf("prove replica update (2) failed: %w", err)})
+
+	}
+	return ctx.Send(SectorProveReplicaUpdate2{
+		Proof: proof,
+	})
+}
+
+func (m *Sealing) handleSubmitReplicaUpdate(ctx statemachine.Context, sector SectorInfo) error {
+
+	tok, _, err := m.Api.ChainHead(ctx.Context())
+	if err != nil {
+		log.Errorf("handleSubmitReplicaUpdate: api error, not proceeding: %+v", err)
+		return nil
+	}
+
+	if err := checkReplicaUpdate(ctx.Context(), m.maddr, sector, tok, m.Api); err != nil {
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+	// XXX use config for various collateral things
+
+	// cfg, err := m.getConfig()
+	// if err != nil {
+	// 	return xerrors.Errorf("getting config: %w", err)
+	// }
+
+	sl, err := m.Api.StateSectorPartition(ctx.Context(), m.maddr, sector.SectorNumber, tok)
+	if err != nil {
+		log.Errorf("handleSubmitReplicaUpdate: api error, not proceeding: %+v", err)
+		return nil
+	}
+	updateProof, err := sector.SectorType.RegisteredUpdateProof()
+	if err != nil {
+		log.Errorf("failed to get update proof type from seal proof: %+v", err)
+		return ctx.Send(SectorSubmitReplicaUpdateFailed{})
+	}
+	enc := new(bytes.Buffer)
+	params := &miner.ProveReplicaUpdatesParams{
+		Updates: []miner.ReplicaUpdate{
+			{
+				SectorID:           sector.SectorNumber,
+				Deadline:           sl.Deadline,
+				Partition:          sl.Partition,
+				NewSealedSectorCID: *sector.UpdateSealed,
```

Comment:
Check if non-nil (otherwise you can brick the miner with a bad `lotus-miner sectors update-state`)

Reply:
the `checkReplicaUpdate` call above will error if sector.UpdateSealed is nil